engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ConstraintTree.java

```diff
@@ -56,21 +62,27 @@
 
 	private final Type validatedValueType;
 
-	public ConstraintTree(ConstraintDescriptorImpl<A> descriptor, Type validatedValueType) {
-		this( descriptor, validatedValueType, null );
+	private volatile Optional<ConstraintValidator<A, ?>> constraintValidatorForDefaultConstraintValidatorFactory;
+
+	private volatile ConcurrentMap<ConstraintValidatorFactory, Optional<ConstraintValidator<A, ?>>> constraintValidatorPerConstraintValidatorFactory;
+
+	public ConstraintTree(ConstraintValidatorManager constraintValidatorManager, ConstraintDescriptorImpl<A> descriptor, Type validatedValueType) {
+		this( constraintValidatorManager, descriptor, validatedValueType, null );
 	}
 
-	private ConstraintTree(ConstraintDescriptorImpl<A> descriptor, Type validatedValueType, ConstraintTree<?> parent) {
+	private ConstraintTree(ConstraintValidatorManager constraintValidatorManager, ConstraintDescriptorImpl<A> descriptor, Type validatedValueType,
```

Comment:
The `ConstraintValidatorManager` isn't really needed here, it's just passed forth and back.

Reply:
I've just pushed a commit to remove it, it rippled through the entire change. Wondering whether it really isn't needed or you just "forgot" to make use of it here? Either way, feel free to include or drop that commit as needed.

Reply:
Nope, I needed it in the constructor before but you're right it's not needed anymore. Instead of undoing it and as we agreed on the volatile usage, I'll just remove the commit introducing it (it was a separate one and things should be OK).

Reply:
Sounds good, thanks for squashing it in!